### DISCUSSION TITLE
DFSPH factor expression with Akinci boundary

### DISCUSSION THREAD

**Discussion by davidAlgis** [InteractiveComputerGraphics/SPlisHSPlasH#326] [Q&A]
Hello, 

I would like to know where the DFSPH factor comes from if we use the fluid/solid coupling of Akinci et al [2012] (see [function line 1078 of TimeStepDFSPH.cpp](https://github.com/InteractiveComputerGraphics/SPlisHSPlasH/blob/master/SPlisHSPlasH/DFSPH/TimeStepDFSPH.cpp)) : 

```math
\alpha_{f_i} = |\sum_{j} m_{f_j} \nabla W_{{f_i}{f_j}}+\sum_{k} \tilde{m}_{b_k} \nabla W_{{f_i}{b_k}}|^2+ \sum_{j} |m_{f_j}\nabla W_{{f_i}{f_j}}|^2
```
where $f_j$ corresponds to the fluid neighbors and $b_k$ to the solid neighbors. 

Indeed, I understand well the DFSPH factor, in the case where we have only fluid particles, but if I try to extend the reasoning when we add solid particles (and in particular dynamic solid particles) I have :

```math
f^p_{f_i} = - \kappa_{f_i}^v (\sum_j m_{f_j} \nabla W_{{f_i}{f_j}} + \sum_{b_k} \tilde{m}_{b_k} \nabla W_{{f_i}{f_j}})
```

We deduce from the symmetric condition $f^p_{f_i} +\sum_j f^p_{f_j\leftarrow f_i}+\sum_k f^p_{b_k\leftarrow f_i}=0$ and by identification : 

```math
\begin{cases}
f^p_{f_j\leftarrow f_i} =  \kappa_{f_i}^v m_{f_j} \nabla W_{{f_i}{f_j}}  \\
f^p_{b_k\leftarrow f_i} =  \kappa_{f_i}^v \tilde{m}_{b_k} \nabla W_{{f_i}{f_j}}
\end{cases}
```

Then we want to minimize $\frac{D\rho_{f_i}}{Dt}$ : 
```math
\begin{align}
\frac{D\rho_{f_i}}{Dt} &=  \sum_j m_{f_j} \mathbf{v_{{f_i}{f_j}}}\cdot \nabla W_{{f_i}{f_j}} +  \sum_k  \tilde{m}_{b_k} \mathbf{v_{{f_i}{b_k}}}\cdot \nabla W_{{f_i}{b_k}}  \\
&= \Delta t(\sum_j m_{f_j} \left(\frac{\mathbf{f}_{f_i}^p}{\rho_{f_i}}- \frac{\mathbf{f}_{{f_j}\leftarrow {f_i}}^p }{\rho_{f_i}} \right)\cdot \nabla W_{{f_i}{f_j}} + \sum_k \tilde{m}_{b_k} \left(\frac{\mathbf{f}_{f_i}^p}{\rho_{f_i}}- \frac{\mathbf{f}_{{b_k}\leftarrow {f_i}}^p }{\rho_{f_i}} \right)\cdot \nabla W_{{f_i}{f_k}})\\
&= \frac{\Delta t \kappa_i^v}{\rho_{f_i}}((\sum_j m_{f_j}\sum_j m_{f_j} \nabla W_{{f_i}{f_j}} +\sum_k \tilde{m}_{b_k} \nabla W_{{f_i}{b_k}} +m_j \nabla W_{{f_i}{f_j}} )\cdot \nabla W_{{f_i}{f_j}} +\sum_k \tilde{m}_{b_k}(\sum_j m_{f_j} \nabla W_{{f_i}{f_j}} + \sum_k \tilde{m}_{b_k} \nabla W_{{f_i}{b_k}}+ \tilde{m}_{b_k} \nabla W_{{f_i}{b_k}})\cdot \nabla W_{{f_i}{b_k}}) 
\end{align}
```

this last expression has a lot of term compared to the result that gives us the expression of $\alpha_i$ above. To retrieve the result we should remove the term where we have $\sum_{f_j}\sum_{b_k}$ and reverse and the term that results from $`\frac{\mathbf{f}_{{b_k}\leftarrow {f_i}}^p}{\rho_{f_i}}`$. How do you explain this ?

Thanks in advance.

**Comment by janbender** [comment]
I was not able to follow your last step. However, last year I wrote a paper where I derive DFSPH in a different way with a complete derivation of the boundary terms: 

https://animation.rwth-aachen.de/publication/0584/

I hope this helps to understand the boundary handling.

**Reply by davidAlgis** [reply]
Thanks for your answer, I still don't completely understand, how come the $\alpha_i$ value used in the ($`|\sum_{j} m_{f_j} \nabla W_{{f_i}{f_j}}+\sum_{k} \tilde{m}_{b_k} \nabla W_{{f_i}{b_k}}|^2+ \sum_{j} |m_{f_j}\nabla W_{{f_i}{f_j}}|^2`$) code seem to not be the same as the one used in the paper ($`|\sum_{j} m_{f_j} \nabla W_{{f_i}{f_j}}+\sum_{k} \tilde{m}_{b_k} \nabla W_{{f_i}{b_k}}|^2+ \sum_{j} |m_{f_j}\nabla W_{{f_i}{f_j}}|^2 +  m_{f_j} \tilde{m}_{b_k}|\nabla W_{{f_i}{f_j}}|^2`$)?

**Reply by davidAlgis** [reply]
In _Consistent SPH Rigid-Fluid Coupling_ they use $`A_{ii} =|\sum_{j} m_{f_j} \nabla W_{{f_i}{f_j}}+\sum_{k} \tilde{m}_{b_k} \nabla W_{{f_i}{b_k}}|^2+ \sum_{j} |m_{f_j}\nabla W_{{f_i}{f_j}}|^2`$ (equation 32 in section 3.4.2), but this is based on the approximation of static boundary particles, in case of dynamic boundary particles shouldn't it be $`|\sum_{j} m_{f_j} \nabla W_{{f_i}{f_j}}+\sum_{k} \tilde{m}_{b_k} \nabla W_{{f_i}{b_k}}|^2+ \sum_{j} |m_{f_j}\nabla W_{{f_i}{f_j}}|^2 +  m_{f_j} \tilde{m}_{b_k}|\nabla W_{{f_i}{f_j}}|^2`$ ?
